feat(allocator): replace allocator_ap2's Vec with Vec2

Dunqing opened the pull request:
Just replace allocator_ap2's Vec with Vec2, and make some changes to make it compile successfully.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9656?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9656** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9656?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9656?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9655** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9655?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9647** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9647?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9646** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9646?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9656).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented on crates/oxc_isolated_declarations/src/lib.rs:
```diff
@@ -79,7 +79,7 @@ impl<'a> IsolatedDeclarations<'a> {
     /// # Errors
     ///
     /// Returns `Vec<Error>` if any errors were collected during the transformation.
-    pub fn build(mut self, program: &Program<'a>) -> IsolatedDeclarationsReturn<'a> {
```
I haven't found a way to avoid explicitly adding lifetime.

Dunqing commented on crates/oxc_isolated_declarations/src/lib.rs:
```diff
@@ -79,7 +79,7 @@ impl<'a> IsolatedDeclarations<'a> {
     ///
     /// Returns `Vec<Error>` if any errors were collected during the transformation.
-    pub fn build(mut self, program: &Program<'a>) -> IsolatedDeclarationsReturn<'a> {
+    pub fn build(mut self, program: &'a Program<'a>) -> IsolatedDeclarationsReturn<'a> {
```
I haven't found a way to avoid explicitly adding lifetime.

Dunqing commented on crates/oxc_allocator/src/vec.rs:
```diff
@@ -167,44 +165,10 @@ impl<'alloc, T> Vec<'alloc, T> {
-    ///
-    /// [owned slice]: Box
-    #[inline]
-    pub fn into_boxed_slice(self) -> Box<'alloc, [T]> {
-        let inner = ManuallyDrop::into_inner(self.0);
-        let slice = inner.leak();
-        let ptr = NonNull::from(slice);
-        // SAFETY: `ptr` points to a valid slice `[T]`.
-        // `allocator_api2::vec::Vec::leak` consumes the inner `Vec` without dropping it.
-        // Lifetime of returned `Box<'alloc, [T]>` is same as lifetime of consumed `Vec<'alloc, T>`,
-        // so data in the `Box` must be valid for its lifetime.
-        // `Vec` uniquely owned the data, and we have consumed the `Vec`, so the new `Box` has
-        // unique ownership of the data (no aliasing).
-        // `ptr` was created from a `&mut [T]`.
-        unsafe { Box::from_non_null(ptr) }
-    }
```
This method does not work directly after replacing it with `Vec2`. Since there no real use cases, so let's remove it instead first

overlookmotel replied:
Can we do that in a separate PR? As this method is not used, it wouldn't even need to be part of this stack.

Dunqing replied:
> Can we do that in a separate PR? As this method is not used, it wouldn't even need to be part of this stack.

#9735

overlookmotel commented:
Here's a minimal reproduction of the lifetimes problem:

```rs
#![expect(missing_docs)]

use bumpalo::{Bump, collections::Vec};

pub struct Tester<'b> {
    bump: &'b Bump,
}

impl<'b> Tester<'b> {
    pub fn test_vec(input: &Vec<'b, u64>, bump: &'b Bump) -> u64 {
        let vec = Vec::from_iter_in(input.iter(), bump);
        //                          ^^^^^^^^^^^^
        // Error: explicit lifetime required in the type of `input`. lifetime `'b` required.
        Self::sum(vec)
    }

    fn sum(vec: Vec<'b, &u64>) -> u64 {
        vec.iter().copied().sum()
    }
}
```

Some discussion (but no complete conclusion) here: https://github.com/fitzgen/bumpalo/issues/171

overlookmotel commented:
After #9733, the changes to `oxc_isolated_declarations` should no longer be required (I hope!)

Dunqing commented:
> After #9733, the changes to `oxc_isolated_declarations` should no longer be required (I hope!)

Yes! You are right!

Dunqing commented on crates/oxc_allocator/src/string.rs:
```diff
@@ -105,15 +105,20 @@ impl<'alloc> String<'alloc> {
         unsafe {
-            let inner = ManuallyDrop::into_inner(bytes.0);
-            let (ptr, len, capacity, bump) = inner.into_raw_parts_with_alloc();
+            let ptr = bytes.as_mut_ptr();
```
I add this safety comment

Dunqing commented:
> Here's a minimal reproduction of the lifetimes problem:
> 
> ```rust
> use bumpalo::{Bump, collections::Vec};
> 
> struct Test<'b> {
>     _dummy: &'b (),
> }
> 
> impl<'b> Test<'b> {
>     pub fn test_vec(input: &Vec<'b, u64>, bump: &'b Bump) -> u64 {
>         let vec = Vec::from_iter_in(input.iter(), bump);
>         //                          ^^^^^^^^^^^^
>         // Error: explicit lifetime required in the type of `input`. lifetime `'b` required.
>         Self::sum(&vec)
>     }
> 
>     fn sum(vec: &Vec<'b, &u64>) -> u64 {
>         vec.iter().copied().sum()
>     }
> }
> ```
> 
> This fixes it:
> 
> ```diff
> - fn sum(vec: &Vec<'b, &u64>) -> u64 {
> + fn sum(vec: &Vec<'_, &u64>) -> u64 {
> ```
> 
> But I don't think that should be necessary. I _think_ it should compile fine as is.
> 
> Some discussion (but no complete conclusion) here: [fitzgen/bumpalo#171](https://github.com/fitzgen/bumpalo/issues/171)

Does that mean we should align the implementation with std does? Accepting allocator rather than a `bump` lifetime?

overlookmotel reviewed (approved):
I remain unsure whether the `T: 'bump` lifetime bound is required on `vec2::Vec` (and on its various impls).

But now the only place this impacts our code is requiring a `C: 'new_alloc` bound on `CloneIn` for `Vec`, which is unproblematic.

So I think we can merge this now, and resolve the lifetime question later on.

overlookmotel commented:
### Merge activity

* **Mar 14, 1:47 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented:
Note that as well as the positive perf impact on transformer, this PR also has a small negative effect on parser benchmarks.

<img width="771" alt="bench" src="https://github.com/user-attachments/assets/d37f66b2-a7f2-4b57-8646-5f90fba987be" />

I imagine this may be for same reason as saw that same performance effect in #9301.

Dunqing commented:
> We can take a look at parser code and see if we can improve that by reserving initial capacity in `Vec`s where we know they're not going to remain empty.

Let me take a look at this

Dunqing committed:
feat(allocator): replace allocator_ap2's Vec with Vec2 (#9656)

Just replace allocator_ap2's Vec with Vec2, and make some changes to make it compile successfully.

overlookmotel commented:
> Let me take a look at this

We may want to look at implementing the optimization in #9301 first. (or that may not even be an optimization any more!)
